File: LinkedLists/328 Odd Even Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        newOddHead = ListNode(0)
        newEvenHead = ListNode(0)
        temp1 = newOddHead
        temp2 = newEvenHead
        isOdd = True
        while(head):
            if isOdd:
                newOddHead.next = head
                newOddHead = newOddHead.next
            else:
                newEvenHead.next = head
                newEvenHead = newEvenHead.next
            head = head.next
            isOdd = not isOdd
        newEvenHead.next = None
        newOddHead.next = temp2.next
        return temp1.next

File: LinkedLists/328 Odd Even Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) 
    {
        ListNode* EvenHead = new ListNode(0);
        ListNode* OddHead = new ListNode(0);
        ListNode* evenNode = EvenHead;
        ListNode* oddNode = OddHead;
        bool isEven = true;
        while(head)
        {
            if(isEven)
            {
                isEven = false;
                EvenHead->next = head;
                EvenHead = EvenHead->next;
            }
            else
            {
                isEven = true;
                OddHead->next = head;
                OddHead = OddHead->next;
            }
            head = head->next;
        }
        OddHead->next = NULL;
        EvenHead->next = oddNode->next;
        return evenNode->next;
    }
};
